utils/upload.go

```diff
@@ -0,0 +1,180 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+// channel to extract files from the folder
+type fileWalk chan string
+
+type Uploader interface {
+	Upload(walker fileWalk)
+}
+
+func UploadToCloudStorage(uploader Uploader, path string) {
+	walker := make(fileWalk)
+	go func() {
+		//get files to upload via the channel
+		if err := filepath.Walk(path, walker.WalkFunc); err != nil {
+			log.Fatalln("Walk failed: ", err)
+		}
+
+		close(walker)
+
+	}()
+
+	uploader.Upload(walker)
+
+}
+
+func (f fileWalk) WalkFunc(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		f <- path
+	}
+
+	return nil
+}
+
+type AwsUploader struct {
+	BucketName      string
+	Prefix          string
+	Region          string
+	FolderLocalPath string
+}
+
+func (a AwsUploader) Upload(walker fileWalk) {
+
+	//creating a new session
+	sess, err := session.NewSession(&aws.Config{
+		Region:                        aws.String(a.Region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+	})
```

Comment:
Here it creates a new session for every request, which is not scalable.

You can take the session as an argument in the function itself, So that the same session can be used many times.
Checkout [GeneratePresignedURL](https://github.com/ZeStream/zestream-server/blob/dev/controllers/upload.go#L14) function for reference

And I think same thing is being done in GCP and Azure as well, So you need to make changes there also

Reply:
okay will make the changes